cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.nus.cool.core.cohort.refactor.olapSelect;
+
+import com.nus.cool.core.cohort.refactor.OlapQueryLayout.granularityType;
+import com.nus.cool.core.cohort.refactor.aggregate.AggregateFactory;
+import com.nus.cool.core.cohort.refactor.aggregate.AggregateFunc;
+import com.nus.cool.core.cohort.refactor.aggregate.AggregateType;
+import com.nus.cool.core.cohort.refactor.storage.OlapRet;
+import com.nus.cool.core.cohort.refactor.storage.ProjectedTuple;
+import com.nus.cool.core.cohort.refactor.storage.RetUnit;
+import com.nus.cool.core.iceberg.result.AggregatorResult;
+import com.nus.cool.core.iceberg.result.BaseResult;
+import com.nus.cool.core.io.readstore.ChunkRS;
+import com.nus.cool.core.io.readstore.FieldRS;
+import com.nus.cool.core.io.readstore.MetaChunkRS;
+import com.nus.cool.core.io.readstore.MetaFieldRS;
+import com.nus.cool.core.io.storevector.InputVector;
+import com.nus.cool.core.schema.FieldType;
+import com.nus.cool.core.util.converter.DayIntConverter;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.map.HashedMap;
+
+public class olapAggregation {
+
+  public enum GroupType {
+
+    STRING,
+
+    NUMERIC
+  }
+
+  private String timeRange;
+
+  private Map<String, BitSet> group = new HashMap<>();
+
+  private List<Map<String, BitSet>> groups = new ArrayList<>();
+
+  private ChunkRS dataChunk;
+
+  private MetaChunkRS metaChunk;
+
+  private AggregateFactory aggregateFactory = new AggregateFactory();
+
+  private String getString(FieldRS field, MetaFieldRS metaField, int value, GroupType type) {
+    switch (type) {
+      case STRING: {
+        InputVector key = field.getKeyVector();
+        return metaField.getString(key.get(value)); // localID => globalId => the filed => string
+      }
+      case NUMERIC: {
+        return String.valueOf(value);
+      }
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+
+  /**
```

Comment:
Use a separate class for groupBy logic. Respect the Single responsibility principle gives us much cleaner codes. It is very hard to understand the code and where we can extend in the future for other types of processing now.

Reply:
Updated